File: finder4j-backtrack-examples/src/main/java/com/github/k0zka/finder4j/backtrack/examples/sudoku/SudokuState.java
package com.github.k0zka.finder4j.backtrack.examples.sudoku;

import com.github.k0zka.finder4j.backtrack.State;

@SuppressWarnings("serial")
public class SudokuState implements State {

	public SudokuState(Short[][] sudoku) {
		super();
		if (sudoku.length != sudokuRowSize && sudoku[0].length != sudokuRowSize) {
			throw new IllegalArgumentException("Requires " + sudokuRowSize
					+ " * " + sudokuRowSize + " array");
		}
		this.sudoku = sudoku;
	}

	final static short sudokuGridSize = 3;
	final static short sudokuRowSize = sudokuGridSize * sudokuGridSize;

	private final Short[][] sudoku;

	public boolean isComplete() {
		for(Short[] row : sudoku) {
			for(Short num : row) {
				if(num == null) {
					return false;
				}
			}
		}
		return true;
	}

	Short[][] getSudoku() {
		return sudoku;
	}

	@Override
	public String toString() {
		StringBuilder str = new StringBuilder();
		for(Short[] row : sudoku) {
			for(Short num : row) {
				str.append(num == null ? "?" : num);
				str.append(' ');
			}
			str.append('\n');
		}
		return str.toString();
	}

}

File: finder4j-backtrack-examples/src/main/kotlin/com/github/k0zka/finder4j/backtrack/examples/sudoku/SudokuState.kt
package com.github.k0zka.finder4j.backtrack.examples.sudoku

import com.github.k0zka.finder4j.backtrack.State

class SudokuState(internal val sudoku: Array<ShortArray>) : State {

	override val isComplete: Boolean
		get() {
			for (row in sudoku) {
				for (num in row) {
					if (num == notSet) {
						return false
					}
				}
			}
			return true
		}

	init {
		if (sudoku.size != sudokuRowSize.toInt() && sudoku[0].size != sudokuRowSize.toInt()) {
			throw IllegalArgumentException("Requires " + sudokuRowSize
												   + " * " + sudokuRowSize + " array")
		}
	}

	override fun toString(): String {
		val str = StringBuilder()
		for (row in sudoku) {
			for (num in row) {
				str.append(num)
				str.append(' ')
			}
			str.append('\n')
		}
		return str.toString()
	}

	companion object {
		val notSet = (-1).toShort()
		internal val sudokuGridSize: Short = 3
		internal val sudokuRowSize = (sudokuGridSize * sudokuGridSize).toShort()
	}

}
